controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -93,46 +89,56 @@ public void refresh() {
     }
 
     @Override
-    CompletableFuture<Void> checkStreamExists(Create create) {
-        CompletableFuture<Void> result = new CompletableFuture<>();
+    CompletableFuture<CreateStreamResponse> checkStreamExists(StreamConfiguration configuration, long timestamp) {
+        CompletableFuture<CreateStreamResponse> result = new CompletableFuture<>();
 
         long time = creationTime.get();
-        if (time != Long.MIN_VALUE && create.getCreationTime() != time) {
-            result.completeExceptionally(StoreException.create(StoreException.Type.DATA_EXISTS, getName()));
-        } else {
-            result.complete(null);
+        synchronized (lock) {
+            if (time != Long.MIN_VALUE) {
+                if (this.configuration != null) {
+                    if (this.state != null) {
+                        State stateVal = (State) SerializationUtils.deserialize(this.state.getData());
+                        if (stateVal.equals(State.UNKNOWN) || stateVal.equals(State.CREATING)) {
+                            CreateStreamResponse.CreateStatus status;
+                            status = (time == timestamp) ? CreateStreamResponse.CreateStatus.NEW :
+                                    CreateStreamResponse.CreateStatus.EXISTS_CREATING;
+                            result.complete(new CreateStreamResponse(status, this.configuration, time));
+                        } else {
+                            result.complete(new CreateStreamResponse(CreateStreamResponse.CreateStatus.EXISTS_ACTIVE, this.configuration, time));
+                        }
+                    } else {
+                        CreateStreamResponse.CreateStatus status = (time == timestamp) ? CreateStreamResponse.CreateStatus.NEW :
+                                CreateStreamResponse.CreateStatus.EXISTS_CREATING;
+
+                        result.complete(new CreateStreamResponse(status, this.configuration, time));
+                    }
+                } else {
+                    result.complete(new CreateStreamResponse(CreateStreamResponse.CreateStatus.NEW, configuration, time));
+                }
+            } else {
+                result.complete(new CreateStreamResponse(CreateStreamResponse.CreateStatus.NEW, configuration, timestamp));
+            }
         }
 
         return result;
     }
 
     @Override
-    CompletableFuture<Void> storeCreationTime(Create create) {
-        Preconditions.checkNotNull(create);
-        Preconditions.checkNotNull(create.getCreationTime());
-        creationTime.compareAndSet(Long.MIN_VALUE, create.getCreationTime());
-        if (creationTime.get() != create.getCreationTime()) {
-            return FutureHelpers.failedFuture(StoreException.create(StoreException.Type.DATA_EXISTS, "createTime"));
-        }
+    CompletableFuture<Void> storeCreationTimeIfNotExists(long timestamp) {
```

Comment:
there is no update api for creation time. its only stored once and never changed.
I can replace the creationTime from atomiclong to a Long which will only be set once if it was null earlier and will do it under lock. Would that be better?